File: src/main/java/com/sequenceiq/cloudbreak/controller/json/InstanceGroupJson.java
package com.sequenceiq.cloudbreak.controller.json;

import java.util.HashSet;
import java.util.Set;

import javax.validation.constraints.Digits;
import javax.validation.constraints.Max;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.sequenceiq.cloudbreak.domain.InstanceGroupType;
import com.wordnik.swagger.annotations.ApiModel;
import com.wordnik.swagger.annotations.ApiModelProperty;

@ApiModel("InstanceGroup")
@JsonIgnoreProperties(ignoreUnknown = true)
public class InstanceGroupJson implements JsonEntity {

    private Long id;
    @NotNull
    @ApiModelProperty(required = true)
    private Long templateId;
    @Min(value = 1, message = "The node count has to be greater than 0")
    @Max(value = 100000, message = "The node count has to be less than 100000")
    @Digits(fraction = 0, integer = 10, message = "The node count has to be a number")
    @ApiModelProperty(required = true)
    private int nodeCount;
    @NotNull
    @ApiModelProperty(required = true)
    private String group;
    @ApiModelProperty(required = true)
    private InstanceGroupType type = InstanceGroupType.HOSTGROUP;
    private Set<InstanceMetaDataJson> metadata = new HashSet<>();

    public InstanceGroupJson() {

    }

    @JsonProperty("id")
    public Long getId() {
        return id;
    }

    @JsonIgnore
    public void setId(Long id) {
        this.id = id;
    }

    public Long getTemplateId() {
        return templateId;
    }

    public void setTemplateId(Long templateId) {
        this.templateId = templateId;
    }

    public int getNodeCount() {
        return nodeCount;
    }

    public void setNodeCount(int nodeCount) {
        this.nodeCount = nodeCount;
    }

    public String getGroup() {
        return group;
    }

    public void setGroup(String group) {
        this.group = group;
    }

    @JsonProperty("metadata")
    public Set<InstanceMetaDataJson> getMetadata() {
        return metadata;
    }

    @JsonIgnore
    public void setMetadata(Set<InstanceMetaDataJson> metadata) {
        this.metadata = metadata;
    }

    public InstanceGroupType getType() {
        return type;
    }

    public void setType(InstanceGroupType type) {
        this.type = type;
    }
}

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/model/InstanceGroupJson.kt
package com.sequenceiq.cloudbreak.api.model

import java.util.HashSet

import javax.validation.constraints.Digits
import javax.validation.constraints.Max
import javax.validation.constraints.Min
import javax.validation.constraints.NotNull

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import com.sequenceiq.cloudbreak.doc.ModelDescriptions
import com.sequenceiq.cloudbreak.doc.ModelDescriptions.InstanceGroupModelDescription

import io.swagger.annotations.ApiModel
import io.swagger.annotations.ApiModelProperty

@ApiModel("InstanceGroup")
@JsonIgnoreProperties(ignoreUnknown = true)
class InstanceGroupJson : JsonEntity {

    @ApiModelProperty(ModelDescriptions.ID)
    var id: Long? = null
    @NotNull
    @ApiModelProperty(value = InstanceGroupModelDescription.TEMPLATE_ID, required = true)
    var templateId: Long? = null
    @Min(value = 1, message = "The node count has to be greater than 0")
    @Max(value = 100000, message = "The node count has to be less than 100000")
    @Digits(fraction = 0, integer = 10, message = "The node count has to be a number")
    @ApiModelProperty(value = InstanceGroupModelDescription.NODE_COUNT, required = true)
    var nodeCount: Int = 0
    @NotNull
    @ApiModelProperty(value = InstanceGroupModelDescription.INSTANCE_GROUP_NAME, required = true)
    var group: String? = null
    @ApiModelProperty(InstanceGroupModelDescription.INSTANCE_GROUP_TYPE)
    var type = InstanceGroupType.CORE
    var metadata: Set<InstanceMetaDataJson> = HashSet()
}
